### DISCUSSION TITLE
Outputting a material property as a vector

### DISCUSSION THREAD

**Discussion by TheGreatCid** [idaholab/moose#28610] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hi all,

I am trying to output a material property as a vector to store all of the quadrature point values on an element. 
I want to output this to Exodus. 

It is unclear to me what property type template I should use for the material property. I cannot find examples of a vector or moosearray type material property. 

Any insight would be appreciated; thank you.

**Comment by GiudGiud** [comment]
Hello

~If you want to visualize it as a vector in paraview, just make sure you output each component, using for example a MaterialRealVectorAUx to 3 variables named with the _x _y and _z suffix

For example:
aaa_x., aaa_y and aaa_z~

**Reply by TheGreatCid** [reply]
So creating 4 properties for each _qp would essentially give the same result? Assuming they are named properly

**Reply by GiudGiud** [reply]
oh nevermind I misread

**Reply by TheGreatCid** [reply]
Also, I did find vector material properties, but I am unsure of how to initialize/output them properly

**Reply by GiudGiud** [reply]
what type is your material property right now? Is it a vector (RealVectorValue)? or Real

**Comment by GiudGiud** [comment]
> I am trying to output a material property as a vector to store all of the quadrature point values on an element.
I want to output this to Exodus.

what for? Exodus is for visualization. If you want to visualize a material property, you project it onto a variable using an auxkernel, such as the MaterialRealAux

If you want to look at qp values, this object will output them to CSV
https://mooseframework.inl.gov/source/vectorpostprocessors/ElementMaterialSampler.html

**Reply by TheGreatCid** [reply]
I am working on a project where I need the quadrature point information on each element for certain material properties. They need to be part of the data in the exodus file.

**Reply by TheGreatCid** [reply]
I can just make a material property for each qp for each element

**Reply by GiudGiud** [reply, marked as answer]
You could make a custom auxkernel, modifying the MaterialRealVectorAux to output only 1 _qp at a time to each variable

**Reply by GiudGiud** [reply]
One variable for each _qp index on every element

**Reply by TheGreatCid** [reply]
That's a good idea as well, thanks!

**Reply by TheGreatCid** [reply]
Well actually, MaterialRealVectorAux already has the option to specify the quad point
